app/components/Showcases/Barter.jsx

```diff
@@ -0,0 +1,724 @@
+                        this.setState(
+                            {from_barter: _from_barter},
+                            this._updateFee
+                        );
+                    } else {
+                        _from_barter[index].from_feeAmount = fee;
+                        _from_barter[index].from_fee_asset_id = fee.asset_id;
+                        this.setState({
+                            from_barter: _from_barter,
+                            from_hasPoolBalance: hasPoolBalance,
+                            from_error: !hasBalance || !hasPoolBalance
+                        });
+                    }
+                } else {
+                    if (should) {
+                        _to_barter[index].to_fee_asset_id = asset_id;
+                        this.setState({to_barter: _to_barter}, this._updateFee);
+                    } else {
+                        _to_barter[index].to_feeAmount = fee;
+                        _to_barter[index].to_fee_asset_id = fee.asset_id;
+                        this.setState({
+                            to_barter: _to_barter,
+                            to_hasPoolBalance: hasPoolBalance,
+                            to_error: !hasBalance || !hasPoolBalance
+                        });
+                    }
+                }
+            })
+        );
+    }
+
+    _getAvailableAssets(state = this.state) {
+        const {from_account, from_error, to_account, to_error} = state;
+
+        let getAssetTypes = (account, err) => {
+            let asset_types = [],
+                fee_asset_types = [];
+            if (!(account && account.get("balances") && !err)) {
+                return {asset_types, fee_asset_types};
+            }
+            let account_balances = account.get("balances").toJS();
+            asset_types = Object.keys(account_balances).sort(utils.sortID);
+            fee_asset_types = Object.keys(account_balances).sort(utils.sortID);
+
+            for (let key in account_balances) {
+                let balanceObject = ChainStore.getObject(account_balances[key]);
+                if (balanceObject && balanceObject.get("balance") === 0) {
+                    asset_types.splice(asset_types.indexOf(key), 1);
+                    if (fee_asset_types.indexOf(key) !== -1) {
+                        fee_asset_types.splice(fee_asset_types.indexOf(key), 1);
+                    }
+                }
+            }
+
+            return {asset_types, fee_asset_types};
+        };
+
+        let from = getAssetTypes(from_account, from_error);
+        let to = getAssetTypes(to_account, to_error);
+
+        return {
+            from_asset_types: from.asset_types || [],
+            to_asset_types: to.asset_types || [],
+            from_fee_asset_types: from.asset_fee_types || [],
+            to_fee_asset_types: to.fee_asset_types || []
+        };
+    }
+
+    addFromAmount() {
+        this.state.from_barter.push("");
```

Comment:
How does this item convert to an actual object with the same struct as the initial one? Confusing

Reply:
object strusture is created in onFromamountChanged method. I will add base structure to this method